File: Protocol/_theoreme_sim/src/generation/chargeInitializer.js
/* Purpose: Initialize charges in a near-empty state so charge growth happens during simulation rounds. */

import { THEOREM_ROOT_ID } from "../core/constants.js";
import { bfsDistances } from "../core/graphUtils.js";

/**
 * @param {Array<{id:number,x:number,y:number}>} nodes
 * @param {Map<number, Set<number>>} adjacency
 * @param {{qForward:number,rootSourceCharge:number}} config
 * @param {() => number} _rng
 * @returns {{nodes:Array<{id:number,x:number,y:number,qTotal:number,isGateway:boolean,eligible:boolean,parent:null,children:number[]}>, distances:Map<number, number>}}
 */
export function initializeCharges(nodes, adjacency, config, _rng) {
  const distances = bfsDistances(adjacency, THEOREM_ROOT_ID);

  const initialized = nodes.map((node) => {
    const qTotal = node.id === THEOREM_ROOT_ID ? config.rootSourceCharge : 0;

    return {
      ...node,
      qTotal,
      isGateway: node.id === THEOREM_ROOT_ID,
      eligible: false,
      parent: null,
      children: [],
    };
  });

  for (const node of initialized) {
    node.eligible = node.qTotal >= config.qForward;
  }

  return { nodes: initialized, distances };
}

File: Protocol/_theoreme_sim_py/src/generation/charge_initializer.py
"""Purpose: Initialize near-empty charge field so dynamics grow during simulation rounds."""

from __future__ import annotations

from src.core.constants import THEOREM_ROOT_ID
from src.core.graph_utils import bfs_distances
from src.core.types import SimNode


def initialize_charges(
    nodes: list[dict],
    adjacency: dict[int, set[int]],
    config: dict,
    _rng,
) -> dict:
    distances = bfs_distances(adjacency, THEOREM_ROOT_ID)

    initialized: list[SimNode] = []
    for node in nodes:
        q_total = float(config["rootSourceCharge"]) if node["id"] == THEOREM_ROOT_ID else 0.0
        initialized.append(
            SimNode(
                id=node["id"],
                x=float(node["x"]),
                y=float(node["y"]),
                q_total=q_total,
                is_gateway=node["id"] == THEOREM_ROOT_ID,
                eligible=False,
                parent=None,
            )
        )

    q_forward = float(config["qForward"])
    for node in initialized:
        node.eligible = node.q_total >= q_forward

    return {
        "nodes": initialized,
        "distances": distances,
    }
